File: android/protocolo/src/main/kotlin/com/marcmayol/dracpaste/protocolo/mensajes/Mensajes.kt
package com.marcmayol.dracpaste.protocolo.mensajes

import com.marcmayol.dracpaste.protocolo.Protocolo
import com.marcmayol.dracpaste.protocolo.cripto.Cripto
import com.marcmayol.dracpaste.protocolo.cripto.aHex
import com.marcmayol.dracpaste.protocolo.red.ProtocoloException
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.util.Base64

/**
 * Los mensajes del protocolo (`docs/protocol.md` §3, §4 y §5) y su ida y vuelta a JSON.
 *
 * El campo `t` identifica el tipo. Se decodifica en dos pasos —primero se mira `t`,
 * después se decodifica la clase concreta— en vez de usar polimorfismo de
 * kotlinx.serialization, porque el discriminador tiene que ser exactamente `t` en el
 * cable y un tipo desconocido no puede reventar: v1 debe poder ignorar en silencio lo
 * que le mande una versión más nueva.
 */
sealed interface Mensaje {
    val t: String
}

// -------------------------------------------------------------- Emparejamiento

@Serializable
data class PairRequest(
    override val t: String = TIPO,
    val v: Int = Protocolo.VERSION,
    val pk: String,
    @SerialName("device_id") val deviceId: String,
    val name: String,
    val token: String,
) : Mensaje {
    companion object { const val TIPO = "PAIR_REQUEST" }
}

@Serializable
data class PairConfirm(
    override val t: String = TIPO,
    @SerialName("device_id") val deviceId: String,
    val name: String,
    val fingerprint: String,
) : Mensaje {
    companion object { const val TIPO = "PAIR_CONFIRM" }
}

@Serializable
data class PairAck(
    override val t: String = TIPO,
    val fingerprint: String,
) : Mensaje {
    companion object { const val TIPO = "PAIR_ACK" }
}

// ------------------------------------------------------------------ Handshake

@Serializable
data class Hello(
    override val t: String = TIPO,
    val v: Int = Protocolo.VERSION,
    @SerialName("device_id") val deviceId: String,
    val challenge: String,
) : Mensaje {
    companion object { const val TIPO = "HELLO" }
}

@Serializable
data class ServerHello(
    override val t: String = TIPO,
    val v: Int = Protocolo.VERSION,
    @SerialName("device_id") val deviceId: String,
    val challenge: String,
) : Mensaje {
    companion object { const val TIPO = "SERVER_HELLO" }
}

@Serializable
data class Auth(
    override val t: String = TIPO,
    val echo: String,
) : Mensaje {
    companion object { const val TIPO = "AUTH" }
}

@Serializable
data class AuthOk(
    override val t: String = TIPO,
    val echo: String,
) : Mensaje {
    companion object { const val TIPO = "AUTH_OK" }
}

// --------------------------------------------------------------------- Sesión

@Serializable
data class Clip(
    override val t: String = TIPO,
    val type: String = Protocolo.TIPO_TEXTO,
    val payload: String,
    @SerialName("timestamp_ms") val timestampMs: Long,
    @SerialName("origin_id") val originId: String,
) : Mensaje {
    companion object {
        const val TIPO = "CLIP"

        /**
         * Construye un CLIP de texto calculando su `origin_id`.
         *
         * @throws ProtocoloException si el texto está vacío o pasa del máximo. Se
         *   comprueba aquí, en el único sitio por el que se crean los clips, y no en
         *   cada llamador.
         */
        fun deTexto(texto: String, timestampMs: Long = System.currentTimeMillis()): Clip {
            val bytes = texto.toByteArray(Charsets.UTF_8)
            if (bytes.isEmpty()) {
                throw ProtocoloException("No se envían clips vacíos")
            }
            if (bytes.size > Protocolo.MAX_CLIP_BYTES) {
                throw ProtocoloException(
                    "El clip ocupa ${bytes.size} bytes y el máximo es ${Protocolo.MAX_CLIP_BYTES}",
                )
            }
            return Clip(
                payload = Base64.getEncoder().encodeToString(bytes),
                timestampMs = timestampMs,
                originId = origenDe(texto),
            )
        }

        /** `origin_id` = SHA-256 del texto en UTF-8, truncado a 16 bytes, en hex. */
        fun origenDe(texto: String): String =
            Cripto.sha256(texto.toByteArray(Charsets.UTF_8)).copyOfRange(0, 16).aHex()
    }

    /** El texto del clip. */
    fun texto(): String = String(Base64.getDecoder().decode(payload), Charsets.UTF_8)

    /** ¿Es un tipo que esta versión sabe manejar? */
    fun esTexto(): Boolean = type == Protocolo.TIPO_TEXTO
}

@Serializable
data class Ping(
    override val t: String = TIPO,
    val seq: Long,
) : Mensaje {
    companion object { const val TIPO = "PING" }
}

@Serializable
data class Pong(
    override val t: String = TIPO,
    val seq: Long,
) : Mensaje {
    companion object { const val TIPO = "PONG" }
}

@Serializable
data class Unpair(override val t: String = TIPO) : Mensaje {
    companion object { const val TIPO = "UNPAIR" }
}

@Serializable
data class Bye(override val t: String = TIPO) : Mensaje {
    companion object { const val TIPO = "BYE" }
}

/** Un tipo que esta versión no conoce. No es un error: se ignora. */
data class MensajeDesconocido(override val t: String) : Mensaje

// -------------------------------------------------------------- Codificación

object CodecMensajes {

    private val json = Json {
        ignoreUnknownKeys = true // Una versión futura puede añadir campos; no es motivo para cortar.
        encodeDefaults = true
        explicitNulls = false
    }

    fun codificar(mensaje: Mensaje): ByteArray {
        val texto = when (mensaje) {
            is PairRequest -> json.encodeToString(PairRequest.serializer(), mensaje)
            is PairConfirm -> json.encodeToString(PairConfirm.serializer(), mensaje)
            is PairAck -> json.encodeToString(PairAck.serializer(), mensaje)
            is Hello -> json.encodeToString(Hello.serializer(), mensaje)
            is ServerHello -> json.encodeToString(ServerHello.serializer(), mensaje)
            is Auth -> json.encodeToString(Auth.serializer(), mensaje)
            is AuthOk -> json.encodeToString(AuthOk.serializer(), mensaje)
            is Clip -> json.encodeToString(Clip.serializer(), mensaje)
            is Ping -> json.encodeToString(Ping.serializer(), mensaje)
            is Pong -> json.encodeToString(Pong.serializer(), mensaje)
            is Unpair -> json.encodeToString(Unpair.serializer(), mensaje)
            is Bye -> json.encodeToString(Bye.serializer(), mensaje)
            is MensajeDesconocido ->
                throw ProtocoloException("No se envía un mensaje de tipo desconocido: ${mensaje.t}")
        }
        return texto.toByteArray(Charsets.UTF_8)
    }

    /**
     * Decodifica un mensaje. Un tipo desconocido devuelve [MensajeDesconocido] en vez de
     * lanzar: quien llama decide ignorarlo, que es lo que manda el protocolo.
     */
    fun decodificar(bytes: ByteArray): Mensaje {
        val objeto = try {
            json.parseToJsonElement(String(bytes, Charsets.UTF_8)) as? JsonObject
                ?: throw ProtocoloException("El mensaje no es un objeto JSON")
        } catch (e: Exception) {
            if (e is ProtocoloException) throw e
            throw ProtocoloException("El mensaje no es JSON válido", e)
        }

        val tipo = objeto["t"]?.jsonPrimitive?.content
            ?: throw ProtocoloException("El mensaje no lleva campo 't'")

        return try {
            when (tipo) {
                PairRequest.TIPO -> json.decodeFromJsonElement(PairRequest.serializer(), objeto)
                PairConfirm.TIPO -> json.decodeFromJsonElement(PairConfirm.serializer(), objeto)
                PairAck.TIPO -> json.decodeFromJsonElement(PairAck.serializer(), objeto)
                Hello.TIPO -> json.decodeFromJsonElement(Hello.serializer(), objeto)
                ServerHello.TIPO -> json.decodeFromJsonElement(ServerHello.serializer(), objeto)
                Auth.TIPO -> json.decodeFromJsonElement(Auth.serializer(), objeto)
                AuthOk.TIPO -> json.decodeFromJsonElement(AuthOk.serializer(), objeto)
                Clip.TIPO -> json.decodeFromJsonElement(Clip.serializer(), objeto)
                Ping.TIPO -> json.decodeFromJsonElement(Ping.serializer(), objeto)
                Pong.TIPO -> json.decodeFromJsonElement(Pong.serializer(), objeto)
                Unpair.TIPO -> json.decodeFromJsonElement(Unpair.serializer(), objeto)
                Bye.TIPO -> json.decodeFromJsonElement(Bye.serializer(), objeto)
                else -> MensajeDesconocido(tipo)
            }
        } catch (e: Exception) {
            throw ProtocoloException("El mensaje '$tipo' no tiene la forma esperada", e)
        }
    }
}

File: windows/DracPaste.Protocolo/Mensajes/Mensajes.cs
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Text.Json.Serialization;
using DracPaste.Protocolo.Red;
using DracPaste.Protocolo.Seguridad;

namespace DracPaste.Protocolo.Mensajes;

/// <summary>
/// Los mensajes del protocolo (<c>docs/protocol.md</c> §3, §4 y §5) y su ida y vuelta a JSON.
///
/// El campo <c>t</c> identifica el tipo. Se decodifica en dos pasos —primero se mira
/// <c>t</c>, después se deserializa la clase concreta— en vez de usar el polimorfismo de
/// System.Text.Json, porque el discriminador tiene que ser exactamente <c>t</c> en el
/// cable y un tipo desconocido no puede reventar: v1 debe poder ignorar en silencio lo
/// que le mande una versión más nueva.
/// </summary>
public abstract record Mensaje
{
    [JsonPropertyName("t")]
    public abstract string T { get; init; }
}

// -------------------------------------------------------------- Emparejamiento

public sealed record PairRequest : Mensaje
{
    public const string Tipo = "PAIR_REQUEST";

    [JsonPropertyName("t")]
    public override string T { get; init; } = Tipo;

    [JsonPropertyName("v")]
    public int V { get; init; } = Protocolo.Version;

    [JsonPropertyName("pk")]
    public required string Pk { get; init; }

    [JsonPropertyName("device_id")]
    public required string DeviceId { get; init; }

    [JsonPropertyName("name")]
    public required string Name { get; init; }

    [JsonPropertyName("token")]
    public required string Token { get; init; }
}

public sealed record PairConfirm : Mensaje
{
    public const string Tipo = "PAIR_CONFIRM";

    [JsonPropertyName("t")]
    public override string T { get; init; } = Tipo;

    [JsonPropertyName("device_id")]
    public required string DeviceId { get; init; }

    [JsonPropertyName("name")]
    public required string Name { get; init; }

    [JsonPropertyName("fingerprint")]
    public required string Fingerprint { get; init; }
}

public sealed record PairAck : Mensaje
{
    public const string Tipo = "PAIR_ACK";

    [JsonPropertyName("t")]
    public override string T { get; init; } = Tipo;

    [JsonPropertyName("fingerprint")]
    public required string Fingerprint { get; init; }
}

// ------------------------------------------------------------------ Handshake

public sealed record Hello : Mensaje
{
    public const string Tipo = "HELLO";

    [JsonPropertyName("t")]
    public override string T { get; init; } = Tipo;

    [JsonPropertyName("v")]
    public int V { get; init; } = Protocolo.Version;

    [JsonPropertyName("device_id")]
    public required string DeviceId { get; init; }

    [JsonPropertyName("challenge")]
    public required string Challenge { get; init; }
}

public sealed record ServerHello : Mensaje
{
    public const string Tipo = "SERVER_HELLO";

    [JsonPropertyName("t")]
    public override string T { get; init; } = Tipo;

    [JsonPropertyName("v")]
    public int V { get; init; } = Protocolo.Version;

    [JsonPropertyName("device_id")]
    public required string DeviceId { get; init; }

    [JsonPropertyName("challenge")]
    public required string Challenge { get; init; }
}

public sealed record Auth : Mensaje
{
    public const string Tipo = "AUTH";

    [JsonPropertyName("t")]
    public override string T { get; init; } = Tipo;

    [JsonPropertyName("echo")]
    public required string Echo { get; init; }
}

public sealed record AuthOk : Mensaje
{
    public const string Tipo = "AUTH_OK";

    [JsonPropertyName("t")]
    public override string T { get; init; } = Tipo;

    [JsonPropertyName("echo")]
    public required string Echo { get; init; }
}

// --------------------------------------------------------------------- Sesión

public sealed record Clip : Mensaje
{
    public const string Tipo = "CLIP";

    [JsonPropertyName("t")]
    public override string T { get; init; } = Tipo;

    [JsonPropertyName("type")]
    public string Type { get; init; } = Protocolo.TipoTexto;

    [JsonPropertyName("payload")]
    public required string Payload { get; init; }

    [JsonPropertyName("timestamp_ms")]
    public required long TimestampMs { get; init; }

    [JsonPropertyName("origin_id")]
    public required string OriginId { get; init; }

    /// <summary>
    /// Construye un CLIP de texto calculando su <c>origin_id</c>.
    /// </summary>
    /// <exception cref="ProtocoloException">
    /// Si el texto está vacío o pasa del máximo. Se comprueba aquí, en el único sitio
    /// por el que se crean los clips, y no en cada llamador.
    /// </exception>
    public static Clip DeTexto(string texto, long? timestampMs = null)
    {
        ArgumentNullException.ThrowIfNull(texto);
        var bytes = Encoding.UTF8.GetBytes(texto);
        if (bytes.Length == 0)
        {
            throw new ProtocoloException("No se envían clips vacíos");
        }

        if (bytes.Length > Protocolo.MaxClipBytes)
        {
            throw new ProtocoloException(
                $"El clip ocupa {bytes.Length} bytes y el máximo es {Protocolo.MaxClipBytes}");
        }

        return new Clip
        {
            Payload = Convert.ToBase64String(bytes),
            TimestampMs = timestampMs ?? DateTimeOffset.UtcNow.ToUnixTimeMilliseconds(),
            OriginId = OrigenDe(texto),
        };
    }

    /// <summary><c>origin_id</c> = SHA-256 del texto en UTF-8, truncado a 16 bytes, en hex.</summary>
    public static string OrigenDe(string texto) =>
        Hex.ToHex(Cripto.Sha256(Encoding.UTF8.GetBytes(texto)).AsSpan(0, 16));

    /// <summary>El texto del clip.</summary>
    public string Texto() => Encoding.UTF8.GetString(Convert.FromBase64String(Payload));

    /// <summary>¿Es un tipo que esta versión sabe manejar?</summary>
    public bool EsTexto() => Type == Protocolo.TipoTexto;
}

public sealed record Ping : Mensaje
{
    public const string Tipo = "PING";

    [JsonPropertyName("t")]
    public override string T { get; init; } = Tipo;

    [JsonPropertyName("seq")]
    public required long Seq { get; init; }
}

public sealed record Pong : Mensaje
{
    public const string Tipo = "PONG";

    [JsonPropertyName("t")]
    public override string T { get; init; } = Tipo;

    [JsonPropertyName("seq")]
    public required long Seq { get; init; }
}

public sealed record Unpair : Mensaje
{
    public const string Tipo = "UNPAIR";

    [JsonPropertyName("t")]
    public override string T { get; init; } = Tipo;
}

public sealed record Bye : Mensaje
{
    public const string Tipo = "BYE";

    [JsonPropertyName("t")]
    public override string T { get; init; } = Tipo;
}

/// <summary>Un tipo que esta versión no conoce. No es un error: se ignora.</summary>
public sealed record MensajeDesconocido : Mensaje
{
    [JsonPropertyName("t")]
    public override string T { get; init; } = string.Empty;
}

// -------------------------------------------------------------- Codificación

public static class CodecMensajes
{
    internal static readonly JsonSerializerOptions Opciones = new()
    {
        // Una versión futura puede añadir campos; no es motivo para cortar la conexión.
        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
        PropertyNamingPolicy = null,

        // Sin esto, el codificador por defecto escapa '+' como + y todo lo que no
        // sea ASCII. Los payloads del protocolo van en base64 —donde '+' es un carácter
        // normal— y los nombres de dispositivo llevan acentos: el mensaje seguiría
        // siendo JSON válido, pero abultaría el triple y sería ilegible al depurarlo.
        // "Unsafe" se refiere a incrustar el resultado en HTML, que no es el caso.
        Encoder = System.Text.Encodings.Web.JavaScriptEncoder.UnsafeRelaxedJsonEscaping,
    };

    public static byte[] Codificar(Mensaje mensaje)
    {
        ArgumentNullException.ThrowIfNull(mensaje);
        var texto = mensaje switch
        {
            PairRequest m => JsonSerializer.Serialize(m, Opciones),
            PairConfirm m => JsonSerializer.Serialize(m, Opciones),
            PairAck m => JsonSerializer.Serialize(m, Opciones),
            Hello m => JsonSerializer.Serialize(m, Opciones),
            ServerHello m => JsonSerializer.Serialize(m, Opciones),
            Auth m => JsonSerializer.Serialize(m, Opciones),
            AuthOk m => JsonSerializer.Serialize(m, Opciones),
            Clip m => JsonSerializer.Serialize(m, Opciones),
            Ping m => JsonSerializer.Serialize(m, Opciones),
            Pong m => JsonSerializer.Serialize(m, Opciones),
            Unpair m => JsonSerializer.Serialize(m, Opciones),
            Bye m => JsonSerializer.Serialize(m, Opciones),
            MensajeDesconocido m => throw new ProtocoloException(
                $"No se envía un mensaje de tipo desconocido: {m.T}"),
            _ => throw new ProtocoloException($"Tipo de mensaje no contemplado: {mensaje.GetType().Name}"),
        };

        return Encoding.UTF8.GetBytes(texto);
    }

    /// <summary>
    /// Decodifica un mensaje. Un tipo desconocido devuelve <see cref="MensajeDesconocido"/>
    /// en vez de lanzar: quien llama decide ignorarlo, que es lo que manda el protocolo.
    /// </summary>
    public static Mensaje Decodificar(byte[] bytes)
    {
        ArgumentNullException.ThrowIfNull(bytes);

        JsonObject objeto;
        try
        {
            objeto = JsonNode.Parse(Encoding.UTF8.GetString(bytes)) as JsonObject
                ?? throw new ProtocoloException("El mensaje no es un objeto JSON");
        }
        catch (JsonException e)
        {
            throw new ProtocoloException("El mensaje no es JSON válido", e);
        }

        var tipo = objeto["t"]?.GetValue<string>()
            ?? throw new ProtocoloException("El mensaje no lleva campo 't'");

        try
        {
            Mensaje? mensaje = tipo switch
            {
                PairRequest.Tipo => objeto.Deserialize<PairRequest>(Opciones),
                PairConfirm.Tipo => objeto.Deserialize<PairConfirm>(Opciones),
                PairAck.Tipo => objeto.Deserialize<PairAck>(Opciones),
                Hello.Tipo => objeto.Deserialize<Hello>(Opciones),
                ServerHello.Tipo => objeto.Deserialize<ServerHello>(Opciones),
                Auth.Tipo => objeto.Deserialize<Auth>(Opciones),
                AuthOk.Tipo => objeto.Deserialize<AuthOk>(Opciones),
                Clip.Tipo => objeto.Deserialize<Clip>(Opciones),
                Ping.Tipo => objeto.Deserialize<Ping>(Opciones),
                Pong.Tipo => objeto.Deserialize<Pong>(Opciones),
                Unpair.Tipo => objeto.Deserialize<Unpair>(Opciones),
                Bye.Tipo => objeto.Deserialize<Bye>(Opciones),
                _ => new MensajeDesconocido { T = tipo },
            };

            return mensaje ?? throw new ProtocoloException($"El mensaje '{tipo}' quedó vacío al decodificar");
        }
        catch (Exception e) when (e is JsonException or InvalidOperationException)
        {
            throw new ProtocoloException($"El mensaje '{tipo}' no tiene la forma esperada", e);
        }
    }
}
